Python/Module6_Testing/Hypothesis.md

```diff
@@ -0,0 +1,535 @@
+```
+<!-- #endregion -->
+
+<!-- #region -->
+Here we have specified that the value of `size` in our test *should be able to take on any integer value within* $[0, 10^{10}]$.
+We did this by using the `integers` "strategy" that is provided by Hypothesis: `st.integers(min_value=0, max_value=1E10)`.
+When we execute the resulting test (which can simply be run within a Jupyter cell or via pytest), this will trigger Hypothesis to generate test cases based on this specification;
+by default, Hypothesis will generate 100 test cases - an amount that we can configure - and will execute our test function for each one of them.
+
+```python
+# Running this test once will trigger Hypothesis to
+# generate 100 values based on the description of our data,
+# and it will execute the test using each one of those values
+>>> test_range_length()
+```
+
+With great ease, we were able to replace our pytest-parameterized test, which only very narrowly tested the property at hand, with a much more robust, Hypothesis-driven test.
+This will be a recurring trend: we will generally produce much more robust tests by _describing_ our data with Hypothesis, rather than manually specifying test values.
+
+The rest of this section will be dedicated to learning about the Hypothesis library and how we can leverage it to write powerful tests.
+<!-- #endregion -->
+
+<!-- #region -->
+<div class="alert alert-warning">
+
+**Hypothesis is _very_ effective...**: 
+
+You may be wondering why, in the preceding example, I arbitrarily picked $10^{10}$ as the upper bound to the integer-values to feed to the test.
+Initially, I didn't write the test that way.
+Instead, I wrote the more general test:
+
+```python
+# `size` can be any non-negative integer
+@given(size=st.integers(min_value=0))
+def test_range_length(size):
+    assert len(range(size)) == size
+```
+
+which places no formal upper bound on the integers that Hypothesis will generate.
+However, this test immediately found an issue (I hesitate to call it an outright bug):
+
+```python
+Falsifying example: test_range_length(
+    size=9223372036854775808,
+)
+
+----> 3     assert len(range(size)) == size
+
+OverflowError: Python int too large to convert to C ssize_t
+```
+
+This reveals that the CPython implementation of the built-in `len` function is such that it can only handle non-negative integers smaller than $2^{63}$ (i.e. it will only allocate 64 bits to represent a signed integer - one bit is used to store the sign of the number).
+Hypothesis revealed this by generating the failing test case `size=9223372036854775808`, which is exactly $2^{63}$.
+I did not want this error to distract from what is otherwise merely a simple example, but it is very important to point out.
+
+Hypothesis has a knack for catching these sorts of unexpected edge cases.
+Now we know that `len(range(size)) == size` _does not_ hold for "arbitrary" non-negative integers!
+(I wonder how many of the Python core developers know about this 😄).
```

Comment:
They know!  After we worked out what was happening I wrote https://github.com/python/cpython/pull/17934 so it's even documented now :joy: